### DISCUSSION TITLE
Gatsby 5 Umbrella Discussion

### DISCUSSION THREAD

**Discussion by Khaledgarbaya** [gatsbyjs/gatsby#36609] [Umbrella Discussions]
Welcome to the overarching discussion related to the release of Gatsby 5! This discussion serves as the hub of all things related to the features, deprecations, breaking changes, docs, and testing that will be part of the general availability of Gatsby 5.

🚢 _We plan on releasing Gatsby 5 in Q4 2022._

## How you can help

Testing and providing feedback are the most valuable things you can do at this stage. Also, be sure to follow this discussion so that you know when we ship new canaries of Gatsby 5. Specific feedback we’re looking for:

- Errors you encounter with Gatsby 5
- Tell us what’s hard about migrating to Gatsby 5, and [contribute to the migration guide](https://github.com/gatsbyjs/gatsby/blob/master/docs/docs/reference/release-notes/migrating-from-v4-to-v5.md)
- Join the [`gatsby-5` Discord channel](https://discord.gg/MhfpnT4cNg)

#### Using our Starter

Clone the existing starter [gatsby-starter-slices](https://github.com/gatsbyjs/gatsby-starter-slices)

```shell
cd gatsby-starter-slices
npm install --legacy-peer-deps
gatsby develop
```

**Using an online IDE**

- [Codesandbox](https://githubbox.com/gatsbyjs/gatsby-starter-slices)
- [Stackblitz](https://githubblitz.com/gatsbyjs/gatsby-starter-slices)

#### Using an existing project

```shell
git checkout -b using-gatsby-5-beta
```

Follow the [Gatsby 4 to 5 migration guide](https://gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/).

### Using Partial Hydration (experimental)

Please visit [RFC: Partial Hydration](https://github.com/gatsbyjs/gatsby/discussions/36608) to learn more.

## So, what’s shipping with Gatsby 5?

We’re shipping two major improvements to Gatsby in Gatsby 5:

- [Slices API](https://github.com/gatsbyjs/gatsby/discussions/36339) which allows component-level incremental builds. Most helpful when headers, footers, navigation, or other highly shared component’s content is sourced from a content management system. Note: File system routes are not supported at this time.
- [Partial Hydration](https://gatsby.dev/v5-partial-hydration) which helps Gatsby sites eliminate unnecessary Javascript from being shipped to the browser (smaller bundles, faster visitor experience, better web vitals)

A handful of breaking changes will happen, please see the [Gatsby 4 to 5 migration guide](https://gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/). The update process will be smooth.
	
## Where can I ask questions?

- Here in this discussion!
- Join the [`gatsby-5` Discord channel](https://discord.gg/MhfpnT4cNg)

**Comment by ChrisLaRocque** [comment]
Getting what looks like peer-deps issues when running the `gatsby new` command called out above. 
![Screen Shot 2022-09-14 at 9 59 41 AM](https://user-images.githubusercontent.com/20352796/190175136-bbc1b82f-924e-4343-aaff-652c25ad1d05.png)

**Reply by Khaledgarbaya** [reply]
can you try installing with `npm install  --legacy-peer-deps`

**Reply by ChrisLaRocque** [reply]
If I `cd` into the directory and run `npm install --legacy-peer-deps` the install completes but I'm given this webpack error instead when running `gatsby develop` 
![Screen Shot 2022-09-14 at 10 05 48 AM](https://user-images.githubusercontent.com/20352796/190176507-b67922b9-7d5c-4e27-82aa-6e3ffa80040d.png)

**Reply by pieh** [reply]
Seems like we are hitting https://github.com/npm/cli/issues/4958 here. I am testing now wether we can force npm to use prerelease versions as candidates for peerDeps

**Reply by pieh** [reply]
https://github.com/gatsbyjs/gatsby-partial-hydration-starter/pull/1 should fix the latter problem (when installed with `npm install --legacy-peer-dep`) as currently it does install some over 1 year old version (`0.0.0-fec00a869` concretely, due to it being "later" when sorted than any `0.0.0-experimental` versions - joys of versions constraints :) ).

I don't think we can really fix the regular `npm install` tho

**Comment by olavea** [comment]
🥳 🏴‍☠️

**Comment by axe312ger** [comment]
Somehow, for me `global` is gone 🤔 

`Uncaught ReferenceError: global is not defined`

Caused by https://github.com/donavon/use-event-listener/blob/develop/src/index.js

**Reply by LekoArts** [reply]
Only change that comes to my mind we did and maybe is relevant:

https://github.com/gatsbyjs/gatsby/blob/c6d3bf8d77fe18cc8f102e163b5cf06271a5b008/packages/gatsby/src/utils/webpack.config.js#L229

**Reply by LekoArts** [reply]
Can you check if this also happens on `gatsby@next` without Partial Hydration? So that there's no regression in Gatsby 4

**Reply by axe312ger** [reply]
Hmm, that error occured without partial hydration.

I'll check with latest v4, gimme some time 🙈

**Reply by axe312ger** [reply]
@LekoArts so on v4, this error appears as soon I go from v4.23.1 to v4.24

So it might be related to partial hydration. It happens also when the flag is disabled.

**Comment by MohdSohail** [comment]
:+1:

**Comment by imjoshin** [comment]
The [Slices API](https://github.com/gatsbyjs/gatsby/discussions/36339) 🍕 has been added to the alpha!

**Comment by imjoshin** [comment]
Hey all! We're excited to announce we're now in **Beta**. 🔥 

You can upgrade to the Gatsby 5 Beta by installing `gatsby@next`. 
We've also published our [migration guide](https://v5.gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/) to the gatsbyjs.com docs!

Thank you all for testing so far, and we are looking forward to hearing more about the beta.

**Comment by alvinometric** [comment]
Are there any plans to support ESM ?

**Reply by bdenham** [reply]
Full ESM support is more important to me than anything else in v5.

**Reply by LekoArts** [reply]
Not as part of the initial 5.0 release, but the plan is to support ESM in Gatsby 5.x

**Reply by LekoArts** [reply]
Please see https://github.com/gatsbyjs/gatsby/discussions/31599#discussioncomment-4193909

**Comment by LekoArts** [comment]
A new version of GraphiQL will be available in Gatsby 5! You can use it now by installing `gatsby@next`. To quote the [README](https://github.com/graphql/graphiql/blob/main/packages/graphiql/README.md):

- Full language support of the latest [GraphQL Specification](https://spec.graphql.org/draft/#sec-Language):
- Syntax highlighting
- Intelligent type ahead of fields, arguments, types, and more
- Real-time error highlighting and reporting for queries and variables
- Automatic query and variables completion
- Automatic leaf node insertion for non-scalar fields
- Documentation explorer with search and markdown support
- Persisted state using `localStorage`
- Simple API for adding custom plugins
- **Dark Mode** 🙏 

![graphiql-overview](https://user-images.githubusercontent.com/16143594/196430557-a321a8c3-e5f3-4735-8626-d0895f3a06bd.png)

**Comment by nrandell** [comment]
Is there any work on updating the graphql source toolkit to graphql 16? This is needed for some other graphql source plugins.

**Reply by LekoArts** [reply]
GraphQL Toolkit is not currently being supported. We have plans in the future (Q1 2023) to continue improving the source plugin authoring experience. This might be a revised GraphQL Toolkit or possibly a new set of tools to help plugin authors.

@smthomas is the product owner of that

**Reply by nrandell** [reply]
I've got a fork of this that is working - https://github.com/nrandell/gatsby-graphql-toolkit and will be submitting a PR soon for that.

**Comment by RobinSJames** [comment]
Is support for GraphQL fragments in `gatsby-node.js` on the roadmap for v5? @LekoArts I know you've commented about this in other issues.

**Reply by LekoArts** [reply]
It's not something we implemented and it's also not high on our priority list. It shouldn't be blocked on anything, so if you (or someone else) wants to see this land in a v5.x, feel free to go ahead.

**Comment by r-ichard** [comment]
Hello, 

After trying to update my project to Gatsby 5 I'm running into an issue with `gatsby-transformer-sharp` 
When I run `gatsby develop` everything seems fine bug once I try to build the project I get the following error : 

```
"gatsby-transformer-sharp" threw an error while running the createSchemaCustomization lifecycle:

Class constructor GraphQLList cannot be invoked without 'new'



  TypeError: Class constructor GraphQLList cannot be invoked without 'new'
  
  - customize-schema.js:457 fluidNodeType
    [gatsby-starter-bootstrap-5]/[gatsby-transformer-sharp]/customize-schema.js:457:15
  
  - customize-schema.js:644 createFields
    [gatsby-starter-bootstrap-5]/[gatsby-transformer-sharp]/customize-schema.js:644:21
  
  - customize-schema.js:846 Object.module.exports [as createSchemaCustomization]
    [gatsby-starter-bootstrap-5]/[gatsby-transformer-sharp]/customize-schema.js:846:13
  
  - api-runner-node.js:487 runAPI
    [gatsby-starter-bootstrap-5]/[gatsby]/src/utils/api-runner-node.js:487:22
  
  - api-runner-node.js:638 Promise.catch.decorateEvent.pluginName
    [gatsby-starter-bootstrap-5]/[gatsby]/src/utils/api-runner-node.js:638:13
  
  - debuggability.js:384 Promise._execute
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/debuggability.js:384:9
  
  - promise.js:518 Promise._resolveFromExecutor
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/promise.js:518:18
  
  - promise.js:103 new Promise
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/promise.js:103:10
  
  - api-runner-node.js:636 
    [gatsby-starter-bootstrap-5]/[gatsby]/src/utils/api-runner-node.js:636:16
  
  - util.js:16 tryCatcher
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/util.js:16:23
  
  - reduce.js:166 Object.gotValue
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/reduce.js:166:18
  
  - reduce.js:155 Object.gotAccum
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/reduce.js:155:25
  
  - util.js:16 Object.tryCatcher
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/util.js:16:23
  
  - promise.js:547 Promise._settlePromiseFromHandler
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/promise.js:547:31
  
  - promise.js:604 Promise._settlePromise
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/promise.js:604:18
  
  - promise.js:649 Promise._settlePromise0
    [gatsby-starter-bootstrap-5]/[bluebird]/js/release/promise.js:649:10
  

not finished createSchemaCustomization - 0.095s


```

It seems related to the latest update to GraphQL 16. My project is open source so feel free to pull to reproduce the error. 

### Reproduction Link

https://github.com/r-ichard/gatsby-starter-bootstrap-5/tree/feature/gatsby-5

### Steps to Reproduce

1. Pull repository on branch `feature/gatsby-5`
2. run `yarn install`
3. run `gatsby build`


### Environment

```shell
System:
    OS: Linux 5.15 elementary OS 6.1 Jólnir
    CPU: (8) x64 Intel(R) Core(TM) i7-8565U CPU @ 1.80GHz
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 18.3.0 - ~/.nvm/versions/node/v18.3.0/bin/node
    Yarn: 1.22.11 - /usr/local/bin/yarn
    npm: 8.11.0 - ~/.nvm/versions/node/v18.3.0/bin/npm
  Browsers:
    Chrome: 106.0.5249.61
    Firefox: 104.0
  npmPackages:
    gatsby: ^5.0.0-next.0 => 5.0.0-next.12
    gatsby-plugin-gatsby-cloud: ^4.24.1-next.2 => 4.24.1
    gatsby-plugin-image: ^2.24.0 => 2.24.0
    gatsby-plugin-manifest: ^4.24.0 => 4.24.0
    gatsby-plugin-offline: ^5.24.0 => 5.24.0
    gatsby-plugin-sass: ^5.24.0 => 5.24.0
    gatsby-plugin-sharp: ^4.24.0 => 4.24.0
    gatsby-source-filesystem: ^4.24.0 => 4.24.0
    gatsby-transformer-sharp: ^4.24.0 => 4.24.0
```
Did you experience something similar ?

**Reply by LekoArts** [reply]
You also need to update all your other `gatsby-*` packages, see https://v5.gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/#update-gatsby-related-packages

**Reply by r-ichard** [reply]
Thanks a lot !  I didn't understand fully what the documentation meant. Here is how my package.json dependencies looks now  : 
```
"gatsby": "^5.0.0-next.0",
"gatsby-plugin-gatsby-cloud": "^5.0.0-next.3",
"gatsby-plugin-image": "^3.0.0-next.5",
"gatsby-plugin-manifest": "^5.0.0-next",
"gatsby-plugin-offline": "^5.0.0-next",
"gatsby-plugin-sass": "^5.23.0-next.0",
"gatsby-plugin-sharp": "^5.0.0-next",
"gatsby-source-filesystem": "^5.0.0-next",
"gatsby-transformer-sharp": "^5.0.0-next",
  ```

**Reply by LekoArts** [reply]
@r-ichard I've added an example to that section, hope that helps

**Comment by ZeldOcarina** [comment]
Hi!
Today I've attempted an upgrade but I am stuck with the issue mentioned here:
#26531 
Do you have any suggestions? It was doing fine with v4!

```
Missing onError handler for invocation 'building-schema', error was 'Error: Cannot
create as TypeComposer the following value:
  GraphQLScalarType({ name: "Date", description: "A date string, such as 2007-12-03,
compliant with the ISO 8601 standard for representation of dates and times using the
Gregorian calendar.", specifiedByURL: undefined, serialize: [function String],
parseValue: [function String], parseLiteral: [function parseLiteral], extensions: {  },
astNode: undefined, extensionASTNodes: [] }).
```

Current package.json dependencies:

```
"dependencies": {
    "@googlemaps/js-api-loader": "^1.14.3",
    "@img-comparison-slider/react": "^7.6.0",
    "@material-ui/core": "^4.12.4",
    "@material-ui/icons": "^4.11.3",
    "@popperjs/core": "^2.11.6",
    "@sendgrid/mail": "^7.7.0",
    "axios": "^0.27.2",
    "babel-plugin-styled-components": "^2.0.7",
    "bootstrap": "^5.1.3",
    "chalk": "4.1.2",
    "cors": "^2.8.5",
    "dotenv": "^16.0.2",
    "gatsby-plugin-google-tagmanager": "^5.0.0",
    "gatsby-plugin-image": "^3.0.0",
    "gatsby-plugin-no-sourcemaps": "^5.0.0",
    "gatsby-plugin-offline": "^6.0.0",
    "gatsby-plugin-page-creator": "^5.0.0",
    "gatsby-plugin-react-helmet": "^6.0.0",
    "gatsby-plugin-robots-txt": "^1.7.1",
    "gatsby-plugin-sass": "^6.0.0",
    "gatsby-plugin-sharp": "^5.0.0",
    "gatsby-plugin-sitemap": "^6.0.0",
    "gatsby-plugin-styled-components": "^6.0.0",
    "gatsby-source-airtable": "^2.3.0",
    "gatsby-source-filesystem": "^5.0.0",
    "gatsby-transformer-remark": "^6.0.0",
    "gatsby-transformer-sharp": "^5.0.0",
    "google-libphonenumber": "^3.2.29",
    "graphql": "^16.6.0",
    "graphql-compose": "^9.0.10",
    "html-react-parser": "^3.0.1",
    "install": "^0.13.0",
    "mongoose": "^6.6.0",
    "ngrok": "^4.3.3",
    "normalize.css": "^8.0.1",
    "npm": "^8.17.0",
    "portfinder": "^1.0.32",
    "postscribe": "^2.0.8",
    "prop-types": "^15.7.2",
    "react-gtm-module": "^2.0.11",
    "react-helmet": "^6.1.0",
    "react-icons": "^4.4.0",
    "react-markdown": "^8.0.3",
    "react-phone-input-2": "^2.15.1",
    "react-responsive": "^9.0.0-beta.10",
    "react-spinners": "^0.13.3",
    "remark-gfm": "^3.0.1",
    "remark-html": "^15.0.1",
    "remark-parse": "^10.0.1",
    "styled-components": "^5.3.5",
    "swiper": "^8.3.2",
    "unified": "^10.1.2",
    "uuid": "^8.3.2"
  },
  "peerDependencies": {
    "gatsby": "^5.0.0",
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  },
  "devDependencies": {
    "@babel/core": "^7.18.10",
    "@emotion/babel-plugin": "^11.9.2",
    "@types/react": "^18.0.17",
    "babel-eslint": "^10.1.0",
    "cli-real-favicon": "^0.0.8",
    "copy-webpack-plugin": "^11.0.0",
    "eslint": "^8.21.0",
    "express": "^4.18.1",
    "gatsby": "^4.19.2",
    "nodemon": "^2.0.19",
    "npm-run-all": "^4.1.5",
    "prettier": "^2.7.1",
    "react": "^17.0.0",
    "react-dom": "^17.0.0",
    "react-is": "^18.2.0",
    "sass": "^1.53.0",
    "typescript": "^4.7.4",
    "webpack": "^5.74.0"
  }
```

**Reply by gh640** [reply]
I met the same (or similar) problem after updating Gatby from v4 to v5 (and related packages).

Environment:

```bash
❯ node -v
v18.12.1

❯ npm -v
8.19.3
```

with `gatsby build`:

```text
Cannot create as TypeComposer the following value:
  GraphQLScalarType({ name: "Date", description: "A date string, such as 2007-12-03, compliant with the
 ISO 8601 standard for representation of dates and times using the Gregorian calendar.",
specifiedByURL: undefined, serialize: [function String], parseValue: [function String], parseLiteral:
[function parseLiteral], extensions: {  }, astNode: undefined, extensionASTNodes: [] }).
```

In my case, multiple versions (v16 and v15) of graphql are installed after I just updated Gatsby to v5:

```text
❯ npm ls graphql
myproject@0.1.0 /path/to/the/project/root
├─┬ gatsby-plugin-feed@5.0.0
│ └─┬ gatsby-plugin-utils@4.0.0
│   └── graphql@16.6.0
├─┬ gatsby-plugin-image@3.0.0
│ └─┬ gatsby-plugin-utils@4.0.0
│   └── graphql@16.6.0
├─┬ gatsby-plugin-manifest@5.0.0
│ └─┬ gatsby-plugin-utils@4.0.0
│   └── graphql@16.6.0
├─┬ gatsby-plugin-sharp@5.0.0
│ └─┬ gatsby-plugin-utils@4.0.0
│   └── graphql@16.6.0
├─┬ gatsby-transformer-sharp@5.0.0
│ └─┬ gatsby-plugin-utils@4.0.0
│   └── graphql@16.6.0
└─┬ gatsby@5.0.0
  ├─┬ @graphql-codegen/add@3.2.1
  │ └── graphql@15.8.0
  ├─┬ @graphql-codegen/core@2.6.5
  │ ├─┬ @graphql-tools/schema@9.0.8
  │ │ ├─┬ @graphql-tools/merge@8.3.10
  │ │ │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ │ │ └── graphql@15.8.0 deduped
  │ │ │ └── graphql@15.8.0 deduped
  │ │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ │ └── graphql@15.8.0 deduped
  │ │ └── graphql@15.8.0 deduped
  │ ├─┬ @graphql-tools/utils@9.0.0
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ @graphql-codegen/plugin-helpers@2.7.2
  │ ├─┬ @graphql-tools/utils@8.13.1
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ @graphql-codegen/typescript-operations@2.5.6
  │ ├─┬ @graphql-codegen/visitor-plugin-common@2.13.1
  │ │ ├─┬ @graphql-tools/optimize@1.3.1
  │ │ │ └── graphql@15.8.0 deduped
  │ │ ├─┬ @graphql-tools/relay-operation-optimizer@6.5.10
  │ │ │ ├─┬ @ardatan/relay-compiler@12.0.0
  │ │ │ │ └── graphql@15.8.0 deduped
  │ │ │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ │ │ └── graphql@15.8.0 deduped
  │ │ │ └── graphql@15.8.0 deduped
  │ │ ├─┬ @graphql-tools/utils@8.13.1
  │ │ │ └── graphql@15.8.0 deduped
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ @graphql-codegen/typescript@2.8.1
  │ ├─┬ @graphql-codegen/schema-ast@2.5.1
  │ │ ├─┬ @graphql-tools/utils@8.13.1
  │ │ │ └── graphql@15.8.0 deduped
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ @graphql-tools/code-file-loader@7.3.10
  │ ├─┬ @graphql-tools/graphql-tag-pluck@7.3.10
  │ │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ │ └── graphql@15.8.0 deduped
  │ │ └── graphql@15.8.0 deduped
  │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ @graphql-tools/load@7.8.4
  │ ├─┬ @graphql-tools/utils@9.0.1
  │ │ └── graphql@15.8.0 deduped
  │ └── graphql@15.8.0 deduped
  ├─┬ express-graphql@0.12.0
  │ └── graphql@15.8.0 deduped
  ├─┬ gatsby-plugin-page-creator@5.0.0
  │ └─┬ gatsby-plugin-utils@4.0.0
  │   └── graphql@16.6.0
  ├─┬ gatsby-plugin-utils@4.0.0
  │ └── graphql@16.6.0 deduped
  ├─┬ graphql-compose@9.0.10
  │ └─┬ graphql-type-json@0.3.2
  │   └── graphql@15.8.0 deduped
  ├─┬ graphql-tag@2.12.6
  │ └── graphql@15.8.0 deduped
  └── graphql@16.6.0
```

I have no idea how this is related but I did the following actions:

1. Remove `package-lock.json` and `node_modules/`
2. Run `npm install --legacy-peer-deps`
3. Run `gatsby clean`

After that, only graphql of v16 is installed and the problem has gone.

I hope this helps :)

**Reply by gh640** [reply]
The following issues might help people down the road:

- https://github.com/gatsbyjs/gatsby/issues/23062
- https://github.com/gatsbyjs/gatsby/issues/24215

**Reply by ZeldOcarina** [reply]
Thanks a lot this solved the issue, it was a gatsby-theme so I've had to juggle all the dependencies a bit but it worked at the end! :)

**Comment by JustFly1984** [comment]
I've updated from 4.24.1, and I still see that Head API is not wrapped in Providers. My app uses react-intl, and we have useIntl hook, which requires Provider to be in the tree, but wrapRootElement is not wrapping Head. There has been a promise to fix that in v5, but here we are, and my app is still broken. Way to go Gatsby team! @marvinjude https://github.com/gatsbyjs/gatsby/discussions/35841#discussioncomment-3592756 

I'm seriously considering to use nextjs instead. At least they are backward compatible, if your is build only on hooks.

**Reply by LekoArts** [reply]
Don't know what you want to achieve with blackmailing us about moving to Next.js 🤷 If your use case is not yet working with Gatsby Head API, then use `react-helmet`. We'll address it sooner than later, it didn't make the cut for 5.0, it'll be in 5.x - that's what we said.

So please stop being so condescending and understand that running a project with X amount of users with X * 2 wishes for features will always push project maintainers into a situation where they have to prioritize things. Plus, it's all open source, you can send in a PR.

**Reply by ZeldOcarina** [reply]
Yeah I was gonna say the same, I also have a very complex setup with `react-helmet` that controls half of the website via airtable, for now it's just not worth it to migrate to the Head API. Sites work anyways.

**Reply by JustFly1984** [reply]
My issue is that I already has had converted the website to Head api from react-helmet, and it was the reason why I found an issue with missing context providers at first place, and keeping the PR open, waiting for wrapRootElement to be implemented.
I'm not blackmailing, I'm complaining.
Please stop making new features, and polish what you already have, to have consistent API.

**Reply by Khaledgarbaya** [reply]
Hi @JustFly1984, your whole attitude is not helping here; I would ask you to please be respectful and have a productive conversation rather than telling us what we should do and what we should not. As we mentioned before this didn't make the cut for 5.0 and will be shipped on 5.x. Meanwhile, feel free to create a detailed issue with the limitations you are hitting, PR are also very welcome if you are interested

**Comment by kevin-singleorigin** [comment]
Hello, 

I am using Gatbsy Cloud. When migrating from v4 to v5 my PR passed the Gatsby Build Service checks (all the feedback and advice above helped me get to that point 😄 ) ; however, after I merged my PR, I encountered the following error: 
> Failed to compile Gatsby files (@parcel/core):
Could not determine version of @parcel/runtime-react-refresh in node_modules/gatsby-parcel-config/lib/index.json. Either include it in "dependencies" or "parcelDependencies"..
File path: /usr/src/app/www/node_modules/gatsby-parcel-config/package.json

I have not seen this error before (and I cannot find other mentions of `parcel` issues), so curious if anyone has advice on things I could try to resolve.

<img width="1535" alt="Screenshot 2022-11-28 at 7 47 33 PM" src="https://user-images.githubusercontent.com/105671752/204434309-623ce225-36ab-4205-b92d-e621c796ced5.png">
<img width="1288" alt="Screenshot 2022-11-28 at 7 47 46 PM" src="https://user-images.githubusercontent.com/105671752/204434311-a3551c3f-fe3f-4703-9252-68aa8eec0b5b.png">


In case it helps, my current `package.json` dependencies (mostly from starting with `gatsby-starter-contentful-homepage`) are:
```
  "dependencies": {
    "@contentful/f36-components": "^4.23.2",
    "@contentful/f36-forms": "^4.14.0",
    "@contentful/f36-tokens": "^4.0.1",
    "@contentful/rich-text-html-renderer": "^15.11.1",
    "@contentful/rich-text-plain-text-renderer": "^15.12.1",
    "@contentful/rich-text-react-renderer": "^15.12.1",
    "@fontsource/dm-mono": "^4.5.4",
    "@fontsource/dm-sans": "^4.5.3",
    "@formspree/react": "^2.3.0",
    "@parcel/core": "^2.8.0",
    "@parcel/runtime-browser-hmr": "^2.8.0",
    "@parcel/runtime-service-worker": "^2.8.0",
    "@vanilla-extract/babel-plugin": "^1.1.4",
    "@vanilla-extract/css": "^1.6.8",
    "@vanilla-extract/css-utils": "^0.1.2",
    "@vanilla-extract/webpack-plugin": "^2.1.5",
    "gatsby": "^5.1.0",
    "gatsby-parcel-config": "^1.2.0",
    "gatsby-plugin-google-analytics": "^5.1.0",
    "gatsby-plugin-google-tagmanager": "^5.1.0",
    "gatsby-plugin-image": "^3.1.0",
    "gatsby-plugin-mailchimp": "^5.2.2",
    "gatsby-plugin-manifest": "^5.1.0",
    "gatsby-plugin-react-helmet": "^6.1.0",
    "gatsby-plugin-sharp": "^5.1.0",
    "gatsby-plugin-sitemap": "^6.1.0",
    "gatsby-plugin-vanilla-extract": "^4.0.0",
    "gatsby-source-contentful": "^8.1.0",
    "gatsby-source-filesystem": "^5.1.0",
    "gatsby-transformer-sharp": "^5.1.0",
    "graphql": "^16.6.0",
    "is-absolute-url": "^4.0.1",
    "react": "^18.2.0",
    "react-bootstrap": "^2.5.0",
    "react-dom": "^18.2.0",
    "react-feather": "^2.0.9",
    "react-helmet": "^6.1.0",
    "sharp": "^0.31.2"
  },
  "devDependencies": {
    "chalk": "^5.1.2",
    "contentful-cli": "^1.10.0",
    "contentful-import": "^8.3.2",
    "gatsby-provision-contentful": "^0.0.4",
    "husky": "^8.0.2",
    "inquirer": "^9.1.4",
    "lint-staged": "^13.0.3",
    "prettier": "2.8.0",
    "yargs-parser": "^21.0.0"
  }
}
```
and also:
```
> node -v
v18.11.0
> npm -v
9.1.2
```

**Reply by Khaledgarbaya** [reply]
This a silly suggestion, but can you try deleting the package-lock and node modules and fo a fresh install?

**Reply by LekoArts** [reply]
You should remove these from your package.json, too

```shell
    "@parcel/core": "^2.8.0",
    "@parcel/runtime-browser-hmr": "^2.8.0",
    "@parcel/runtime-service-worker": "^2.8.0",
    "gatsby-parcel-config": "^1.2.0",
    "graphql": "^16.6.0",
    "sharp": "^0.31.2"
```

**Reply by kevin-singleorigin** [reply]
Thanks so much for the advice - I will try these and report back :)

**Reply by kevin-singleorigin** [reply]
I did both of the above, and everything is looking good now :) I am relatively new to Gatsby, so I appreciate the help - you probably saved me hours of fiddling.

**Comment by SevenOutman** [comment]
Got `l.ZP.loadPage(...).then is not a function` error with Gatsby 5.2.0, when I click a Link to a page with ChunkLoadError in production.

<img width="458" alt="image" src="https://user-images.githubusercontent.com/8225666/206972789-a2fef090-7351-46ea-bcd6-bf1213635ce3.png">

The error seems to be thrown from this line in `navgation.js` https://github.com/gatsbyjs/gatsby/blob/dc3d741260e057540ed1294558df78aa63126a8b/packages/gatsby/cache-dir/navigation.js#L90 where the `loader.loadPage(...)` may return something that's not a Promise when page is containing errors, according to `loader.js` https://github.com/gatsbyjs/gatsby/blob/dc3d741260e057540ed1294558df78aa63126a8b/packages/gatsby/cache-dir/loader.js#L349-L354

FYI I'm still trying to figure out why the ChunkLoadError occurs, it looks like rendering a specific component from a third-party library is causing the error.

**Comment by tlylt** [comment]
I am upgrading from Gatsby v4 to v5. Not sure if it's because I'm using windows and nvm, I am getting the following error when running `npx gatsby-codemods@latest sort-and-aggr-graphql .`
> $ npx gatsby-codemods@latest sort-and-aggr-graphql "C:\Users\User\Documents\GitHub\blogging"
Executing command: jscodeshift --ignore-pattern=**/node_modules/** --ignore-pattern=**/.cache/** --ignore-pattern=**/public/** --extensions=jsx,js,ts,tsx --transform C:\Users\User\AppData\Local\npm-cache\_npx\6a274bac855c0dc9\node_modules\gatsby-codemods\transforms\sort-and-aggr-graphql.js C:\Users\User\Documents\GitHub\blogging
C:\Users\User\AppData\Local\npm-cache\_npx\6a274bac855c0dc9\node_modules\.bin\jscodeshift:2
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

>SyntaxError: missing ) after argument list
    at internalCompileFunction (node:internal/vm:74:18)
    at wrapSafe (node:internal/modules/cjs/loader:1141:20)
    at Module._compile (node:internal/modules/cjs/loader:1182:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1272:10)
    at Module.load (node:internal/modules/cjs/loader:1081:32)
    at Module._load (node:internal/modules/cjs/loader:922:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47

> Node.js v18.13.0

**Update**
I managed to get it working via the [jscodeshift](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-codemods#jscodeshift) method.

**Comment by that1matt** [comment]
We are in the process of upgrading from Gatsby 4.25.2 to 5.4.0, but we are seeing an issue with our plugins not loading.

We currently have a monorepo setup like this

```
    main/
        site1/
            package.json
        site2/
            package.json
        gatsby-plugins/
            plugins/
                gatsby-plugin-name/
                    gatsby-node.ts
    package.json
```

We currently have several plugins that do not load on Gatsby v5 if they are using `gatsby-node.ts` inside the `main > gatsby-plugins > plugins > plugin-name > gatsby-node.ts`. If I change the `gatsby-node.ts` to be `gatsby-node.js`, gatsby loads the plugin and the data. The plugin folder will appear in `.cache/caches/plugin-name` (this folder is not created when using the `gatsby-node.ts`) and not in `.cache/complied/plugins/plugin-name` if the plugin lives inside a gatsby site, under the folder `plugins/plugin-name`.

Another Note: If you have a `gatsby-browser.tsx/gatsby-server.tsx` file that lives in the same gatsby-plugin as the `gatsby-node.ts`, it will show these `browserApis/ssrApis` in the graphql output from `allSitePlugin`, but the `nodeApis` is an empty array.

I have been trying to figure out why these are not loading and the difference from v4.25.2 and v5, but I am unable to track down this specific change with as many changes that have happened.

My question, is this a bug in v5+, or is this a bug in v3+/v4+ that got fixed in v5+. I didn't see anything related to this in the breaking changes section on the release notes that related to this type of import. 

I have also created a simple minimal repo here to show the issue. https://github.com/that1matt/gatsby-node-typescript-loading-error

Changing the name of `gatsby-node.bak.js` to `gatsby-node.js` and `gatsby-node.ts` to `gatsby-node.bak.ts` then running a `clean && start` should not create the graphql source nodes.

**Reply by that1matt** [reply]
I have tracked this down some more, and when calling the [resolveJSFilepath](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/resolve-module-exports.ts#L209) it errors out, but doesn't throw an error as the [resolveJSFilepath](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/bootstrap/resolve-js-file-path.ts#L17) returns an empty string.

Obviously the issue stems from only importing `.js` or `.mjs` extensions in this file. 

Still not sure if this is a bug, or something that was intentionally changed. 

Any help would be much appreciated.
